File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        int[] res = new int[k];
        Map<Integer, Integer> count = new HashMap<>();
        ArrayList<ArrayList<Integer>> freq = new ArrayList<>();
        for (int i = 0; i <= nums.length; i++) {
            freq.add(new ArrayList<>());
        }
        for (int n : nums) {
            count.put(n, count.getOrDefault(n, 0) + 1);
        }
        for (int key : count.keySet()) {
            freq.get(count.get(key)).add(key);
        }
        int index = 0;
        for (int i = nums.length; i > 0; i--) {
            if (freq.get(i).isEmpty()) {
                continue;
            } 
            else {
                for (int j : freq.get(i)) {
                    res[index] = j;
                    index++;
                    if (index == k) {
                        return res;
                    }
                }
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-1.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False
        cache = {}

        def memo(i, j, k):
            if k == len(s3):
                return i == len(s1) and j == len(s2)
            if (i, j) in cache:
                return cache[(i, j)]

            res = False
            if i < len(s1) and s1[i] == s3[k]:
                res = memo(i + 1, j, k + 1)
            if not res and j < len(s2) and s2[j] == s3[k]:
                res = memo(i, j + 1, k + 1)
            
            cache[(i, j)] = res
            return res
        
        return memo(0, 0, 0)
                
            
        